src/utils/component-versioning.js

```diff
@@ -0,0 +1,99 @@
+import { html } from 'lit-element';
+import defineOnce from './define-once';
+
+// helper functions
+const toKebabCase = dottedVersionString =>
+  `${dottedVersionString}`.replace(/\./g, '-').replace(/[^A-Za-z0-9-]/g, '');
+
+const versionedTag = (tagName, version) => `${tagName}-${toKebabCase(version)}`;
+
+const extractDependencies = componentClass => {
+  const { versions, tagName } = componentClass;
+  // extract all dependencies by comparing with master-component tag name
+  const dependencies = Object.keys(versions).filter(name => name !== tagName);
+  // pair the list of dependency tag names with their versions
+  return [dependencies, dependencies.map(_tagName => versions[_tagName])];
+};
+
+const oldTag = (tagName, closing = '') => `<${closing}${tagName}`;
+
+const newTag = (tagName, aVersion, closing) =>
+  oldTag(versionedTag(tagName, aVersion), closing);
+
+// Example: someStrings = ['<div><axa-dropdown .items="','" </axa-dropdown></div>']
+//          aTagname = 'axa-dropdown', aVersion = '7.0.2'
+//
+// Rewritten as           ['<div><axa-dropdown-7-0-2 .items="','" </axa-dropdown-7-0-2></div>'].
+//
+// Note: this uses the split-join technique to perform global string substitution
+// without needing the special-character escaping necessary
+// a reg-exp-based alternative (the latter is marginally faster, but our strings here are short)
+const rewrite = (someStrings, aTagName, aVersion) =>
+  someStrings.map(string =>
+    string
+      .split(oldTag(aTagName))
+      .join(newTag(aTagName, aVersion))
+      .split(oldTag(aTagName, '/'))
+      .join(newTag(aTagName, aVersion, '/'))
+  );
+
+// ///
+// API functions
+// ///
+
+// examples: defineVersioned(AXADatepicker, AXADropdown, AXAButton], __VERSION_INFO__);
+//           defineVersioned([AXADatepicker],  __VERSION_INFO__);
+//           defineVersioned([this, AXADropdown],  __VERSION_INFO__);
+//           defineVersioned([AXACheckbox], 'rsv');
+const defineVersioned = (dependencies, versionInfo) => {
+  const customVersion = typeof versionInfo === 'string' && versionInfo;
+  // process all dependant components that it lists...
+  let versionedTagName = '';
+  dependencies.forEach(dependency => {
+    const componentClass =
+      dependency instanceof HTMLElement ? dependency.constructor : dependency;
+    // ordinary, non-POD versioning?
+    if (!customVersion) {
+      // yes, inject version info into component-defining class
+      const { tagName } = componentClass;
+      componentClass.versions = versionInfo[tagName];
```

Comment:
@markus-walther this is super dangerous. Please don't use the the class as a memory/state object, make a memory object away from it instead. What if a dev or new api starts defined `static versions` in the class? Performance wise is negligible the difference.

Reply:
As discussed, I will check for the presence of a pre-existing `versions` class property and throw an error if found.

Rationale for automatic injection of `versions` here was twofold:

1. Why automatic? Remove an error case for developers who might forget to version a component otherwise
1. Why to component class? To aid live debugging using dev tools console, a la `window.customElements.get('axa-datepicker').versions`